src/error/src/tonic.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::error::Error;
+use std::sync::Arc;
+
+#[easy_ext::ext(ToTonicStatus)]
+impl<T> T
+where
+    T: ?Sized + std::error::Error,
+{
+    /// Convert the error to [`tonic::Status`] with the given [`tonic::Code`].
+    ///
+    /// The source chain is preserved by pairing with [`TonicStatusWrapper`].
+    // TODO(error-handling): disallow constructing `tonic::Status` directly with `new` by clippy.
+    pub fn to_status(&self, code: tonic::Code) -> tonic::Status {
+        // Embed the whole error (`self`) and its source chain into the details field.
+        // At the same time, set the message field to the error message of `self` (without source chain).
+        // The redundancy of the current error's message is intentional in case the client ignores the `details` field.
+        let source = serde_error::Error::new(self);
+        let details = bincode::serialize(&source).unwrap_or_default();
+
+        let mut status = tonic::Status::with_details(code, self.to_string(), details.into());
```

Comment:
I'm not sure whether this is idiomatic. 

IIUC, `details` has a specific definition here:
- https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto
- tonic's: https://docs.rs/tonic-types/latest/tonic_types/struct.ErrorDetails.html

We might need to put extra stuff in `metadata` 
- https://docs.rs/tonic-types/latest/tonic_types/trait.StatusExt.html

I'm not sure if the official `details` contains `source` and stuff.

Reply:
Wow, didn't investigate this crate. Seems it's just the generated Rust code for the `ErrorDetails` message of gRPC standard.

> We might need to put extra stuff in `metadata`

Sure, I'll do that. Actually the `details` is also encoded in the metadata header with a reserved key.